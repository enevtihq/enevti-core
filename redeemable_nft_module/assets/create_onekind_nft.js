const { BaseAsset } = require("lisk-sdk");
const { REDEEMABLE_NFT_CREATE_ONEKIND_NFT } = require("../constants/id");
const UTILITY = require("../constants/utility");
const RECURRING = require("../constants/recurring");
const NFTTYPE = require("../constants/nft_type");
const CreateOneKindNFTAssetSchema = require("../schemas/assets/create_onekind_nft");
const { generateID, getAllNFT, setAllNFT, getAllNFTContainer, setAllNFTContainer, setNFTById, setNFTContainerById } = require("../utils/chain_state");
const CHAIN = require("../constants/chain");
const RARITY = require("../constants/rarity");
const JOURNEYACTIVITY = require("../constants/journey_activity");
const REDEEMSTATUS = require("../constants/redeem_status");
const { asyncForEach } = require("../utils/helper");
const CreateOneKindNFTValidator = require("../external-validator/create_onekind_nft");

class CreateOneKindNFTAsset extends BaseAsset {
  name = "createOneKindNFTAsset";
  id = REDEEMABLE_NFT_CREATE_ONEKIND_NFT;
  schema = CreateOneKindNFTAssetSchema;

  validate({ asset }) {
    if (asset.name.length > 18) {
      throw new Error(`asset.name must be below 18 char length`);
    }
    if (!Object.values(UTILITY).includes(asset.utility)) {
      throw new Error(`asset.utility is unknown`);
    }
    if (!Object.values(RECURRING).includes(asset.recurring)) {
      throw new Error(`asset.recurring is unknown`);
    }
    if (asset.recurring === RECURRING.PERWEEK && asset.time.day < 0) {
      throw new Error(`asset.time.day is required on recurring perweek`);
    }
    if (asset.recurring === RECURRING.PERMONTH && asset.time.date <= 0) {
      throw new Error(`asset.time.date is required on recurring permonth`);
    }
    if (asset.recurring === RECURRING.PERYEAR && (asset.time.date <= 0 || asset.time.month <= 0)) {
      throw new Error(`asset.time.date and asset.time.month are required on recurring peryear`);
    }
    if (asset.recurring === RECURRING.ONCE && (asset.time.date <= 0 || asset.time.month <= 0 || asset.time.year <= 0)) {
      throw new Error(`asset.time.date, asset.time.month, and asset.time.year are required on recurring once`);
    }
    if (asset.redeemLimit <= 0) {
      throw new Error(`asset.redeemLimit must be greater than 0`);
    }
    if (asset.price <= 0) {
      throw new Error(`asset.price must be greater than 0`);
    }
    if (asset.quantity <= 0) {
      throw new Error(`asset.quantity must be greater than 0`);
    }
    if (asset.until <= BigInt(0)) {
      throw new Error(`asset.quantity must be greater than 0`);
    }
    if (asset.mintingExpire !== -1 && asset.mintingExpire <= 10000) {
      throw new Error(`valid asset.mintingExpire is in milliseconds and must be greater than 10000 (10 sec)`);
    }
    if (asset.royalty.origin <= 0) {
      throw new Error(`asset.royalty.origin must be greater than 0`);
    }
    if (asset.royalty.staker <= 0) {
      throw new Error(`asset.royalty.staker must be greater than 0`);
    }
  }

  async apply({ asset, stateStore, reducerHandler, transaction }) {
    const senderAddress = transaction.senderAddress;
    const allRegisteredNFT = await getAllNFT(stateStore);
    const timestampInSec = await stateStore.chain.lastBlockHeaders[0].timestamp;

    const externalValidation = CreateOneKindNFTValidator(asset, stateStore, reducerHandler, transaction);
    if (externalValidation.status !== true) {
      throw new Error(`Error during external validation: ${externalValidation.message}`);
    }

    const allNFT = [];
    for (let i = 0; i < asset.quantity; i++) {
      const nft = {
        id: generateID(senderAddress, transaction.nonce + BigInt(i + 1)),
        serial: asset.name + "-" + i.toString(),
        name: asset.name,
        description: asset.description,
        createdOn: timestampInSec,
        data: asset.data,
        ownerAddress: "",
        originAddress: senderAddress,
        originChain: CHAIN.NAME,
        NFTType: NFTTYPE.ONEKIND,
        utility: asset.utility,
        redeem: {
          parts: [],
          partition: [],
          count: 0,
          message: {
            cipher: "",
            nonce: "",
          },
          status: REDEEMSTATUS.IDLE,
          recurring: asset.recurring,
          time: asset.time,
          from: asset.from,
          until: asset.until,
          limit: asset.redeemLimit,
        },
        rarity: RARITY.UNDEFINED,
        journey: [
          {
            activity: JOURNEYACTIVITY.CREATED,
            date: timestampInSec,
            by: senderAddress,
            from: "",
            message: "",
            amount: 0,
          },
        ],
        price: {
          amount: asset.price.amount,
          currency: asset.price.currency,
        },
        onSale: false,
        royalty: {
          origin: asset.royalty.origin,
          staker: asset.royalty.staker,
        },
      };
      allNFT.push(nft);
      allRegisteredNFT.push(nft.id);
    }

    asyncForEach(allNFT, async (item) => {
      await setNFTById(stateStore, item.id, item);
    });

    const NFTContainer = {
      id: generateID(senderAddress, transaction.nonce),
      containerType: NFTTYPE.ONEKIND,
      allNFT: allNFT,
      NFTSold: [],
      availableItem: allNFT,
      mintingExpire: asset.mintingExpire === -1 ? -1 : Math.floor(asset.mintingExpire / 1000),
      originAddress: senderAddress,
      price: {
        amount: asset.price.amount,
        currency: asset.price.currency,
      },
      name: asset.name,
      description: asset.description,
      packSize: 1,
      createdOn: timestampInSec,
    };

    const allNFTContainer = await getAllNFTContainer(stateStore);
    allNFTContainer.push(NFTContainer);

    await setNFTContainerById(stateStore, NFTContainer.id, NFTContainer);
    await setAllNFTContainer(allNFTContainer);
    await setAllNFT(stateStore, allRegisteredNFT);
  }
}

module.exports = CreateOneKindNFTAsset;
