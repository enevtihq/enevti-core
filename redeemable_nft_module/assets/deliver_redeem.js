const { BaseAsset } = require("lisk-sdk");
const { REDEEMABLE_NFT_REQUEST_REDEEM } = require("../constants/id");
const RequestRedeemAssetSchema = require("../schemas/assets/request_redeem");
const REDEEMSTATUS = require("../constants/redeem_status");
const { getNFTById, setNFTById, getRedeemMonitor, setRedeemMonitor } = require("../utils/chain_state");
const JOURNEYACTIVITY = require("../constants/journey_activity");

class RequestRedeemAsset extends BaseAsset {
  name = "requestRedeemAsset";
  id = REDEEMABLE_NFT_REQUEST_REDEEM;
  schema = RequestRedeemAssetSchema;

  async apply({ asset, stateStore, reducerHandler, transaction }) {
    const senderAddress = transaction.senderAddress;
    const senderAccount = await stateStore.account.get(senderAddress);
    const timestampInSec = await stateStore.chain.lastBlockHeaders[0].timestamp;
    const nft = await getNFTById(stateStore, asset.nftId);

    if (!nft) {
      throw new Error("NFT doesn't exist");
    }

    if (!nft.originAddress.equals(senderAddress)) {
      throw new Error("Sender is not the origin of this NFT");
    }

    if (!Object.values(senderAccount.redeemableNFT.creator.requestQueue).includes(nft.id)) {
      throw new Error("NFT is not in origin requesQueue");
    }

    if (nft.redeem.status !== REDEEMSTATUS.REQUESTED) {
      throw new Error("NFT status is not requested");
    }

    nft.redeem.status = REDEEMSTATUS.READY;
    nft.redeem.count++;
    nft.redeem.message.cipher = asset.cipher;
    nft.redeem.message.nonce = asset.nonce;

    nft.journey.unshift({
      activity: JOURNEYACTIVITY.DELIVERED,
      date: timestampInSec,
      by: nft.ownerAddress,
      from: senderAddress,
      message: `Requested by ${nft.ownerAddress.toString("hex")} on: ${new Date(nft.redeem.touched).toUTCString()}`,
      amount: 0,
    });

    nft.redeem.touched = timestampInSec;
    senderAccount.redeemableNFT.creator.delivered++;
    const senderRequestIndex = senderAccount.redeemableNFT.creator.requestQueue.findIndex((a) => a.equals(nft.id));
    senderAccount.redeemableNFT.creator.requestQueue.splice(senderRequestIndex, 1);

    const redeemMonitor = await getRedeemMonitor(stateStore);
    const redeemMonitorItemIndex = redeemMonitor.all.findIndex((a) => {
      a.id.equals(nft.id);
    });
    redeemMonitor.all[redeemMonitorItemIndex] = nft;

    await setNFTById(stateStore, nft.id, nft);
    await stateStore.account.set(senderAddress, senderAccount);
    await setRedeemMonitor(stateStore, redeemMonitor);
  }
}

module.exports = RequestRedeemAsset;
