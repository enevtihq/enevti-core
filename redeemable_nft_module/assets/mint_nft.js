const { BaseAsset } = require("lisk-sdk");
const seedrandom = require("seedrandom");
const { REDEEMABLE_NFT_MINT_NFT } = require("../constants/id");
const JOURNEYACTIVITY = require("../constants/journey_activity");
const NFTTYPE = require("../constants/nft_type");
const MintNFTAssetSchema = require("../schemas/assets/mint_nft");
const { getNFTContainerById, getNFTPackById, setNFTPackById, getNFTById, setNFTById, setNFTContainerById } = require("../utils/chain_state");
const { asyncForEach } = require("../utils/helper");

class MintNFTAsset extends BaseAsset {
  name = "mintNFTAsset";
  id = REDEEMABLE_NFT_MINT_NFT;
  schema = MintNFTAssetSchema;

  validate({ asset }) {
    if (asset.quantity <= 0) {
      throw new Error(`asset.quantity must be greater than 0`);
    }
    if (asset.timestamp > Date.now()) {
      throw new Error(`asset.timestamp can't be in the future`);
    }
  }

  async apply({ asset, stateStore, reducerHandler, transaction }) {
    let container = await getNFTContainerById(stateStore, asset.containerId);
    const senderAddress = transaction.senderAddress;
    const senderAccount = await stateStore.account.get(senderAddress);
    const rng = seedrandom(transaction.nonce);

    if (!container) {
      throw new Error("NFT Container doesn't exist");
    }

    if (container.mintingExpire !== -1) {
      const blockHeight = await stateStore.chain.lastBlockHeaders.height;
      if (container.mintingExpire < blockHeight) {
        throw new Error("minting expired");
      }
    }

    if (container.availableItem.length < asset.quantity) {
      throw new Error("quantity unavailable");
    }

    let boughtItem = [];
    for (let i = 0; i < asset.quantity; i++) {
      switch (container.containerType) {
        case NFTTYPE.ONEKIND:
          boughtItem.unshift(container.availableItem[0]);
          container.NFTSold.push(container.availableItem[0]);
          container.availableItem.shift();
          break;
        case NFTTYPE.UPGRADABLE:
          boughtItem.unshift(container.availableItem[0]);
          container.NFTSold.push(container.availableItem[0]);
          container.availableItem.shift();
          break;
        case NFTTYPE.PACKED:
          const index = Math.floor(rng() * container.availableItem.length);
          const item = container.availableItem[index];
          const itemIndex = container.availableItem.findIndex((t) => t.equals(item));
          let packItem = await getNFTPackById(stateStore, item);
          container.availableItem.splice(itemIndex, 1);
          boughtItem = boughtItem.concat(packItem.NFTItem);
          container.NFTSold = container.NFTSold.concat(packItem.NFTItem);
          packItem.boughtBy = senderAddress;
          await setNFTPackById(stateStore, packItem.id, packItem);
          break;
        default:
          throw new Error("FATAL: unkown container type");
      }
    }

    asyncForEach(boughtItem, (item) => {
      const nft = await getNFTById(stateStore, item);
      nft.ownerAddress = senderAddress;
      nft.journey.unshift({
        activity: JOURNEYACTIVITY.MINTED,
        date: asset.timestamp,
        by: senderAddress,
        from: container.originAddress,
        message: `${container.name} Pack`,
        amount: 0,
      });
      senderAccount.redeemableNFT.ownNFTs.push(nft.id);
      await stateStore.account.set(senderAddress, senderAccount);
      await setNFTById(stateStore, nft.id, nft);
    });

    await setNFTContainerById(stateStore, container.id, container);

    await reducerHandler.invoke("token:debit", {
      address: senderAddress,
      amount: container.price.amount,
    });

    await reducerHandler.invoke("token:credit", {
      address: container.originAddress,
      amount: container.price.amount,
    });
  }
}

module.exports = MintNFTAsset;
