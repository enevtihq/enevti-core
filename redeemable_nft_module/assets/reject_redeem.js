const { BaseAsset } = require("lisk-sdk");
const { REDEEMBALE_NFT_REJECT_REDEEM } = require("../constants/id");
const REDEEMSTATUS = require("../constants/redeem_status");
const { getNFTById, setNFTById, getRedeemMonitor, setRedeemMonitor } = require("../utils/chain_state");
const JOURNEYACTIVITY = require("../constants/journey_activity");
const RejectRedeemAssetSchema = require("../schemas/assets/reject_redeem");

class RejectRedeemAsset extends BaseAsset {
  name = "rejectRedeemAsset";
  id = REDEEMBALE_NFT_REJECT_REDEEM;
  schema = RejectRedeemAssetSchema;

  async apply({ asset, stateStore, reducerHandler, transaction }) {
    const senderAddress = transaction.senderAddress;
    const senderAccount = await stateStore.account.get(senderAddress);
    const timestampInSec = await stateStore.chain.lastBlockHeaders[0].timestamp;
    const nft = await getNFTById(stateStore, asset.nftId);

    if (!nft) {
      throw new Error("NFT doesn't exist");
    }

    if (!nft.originAddress.equals(senderAddress)) {
      throw new Error("Sender is not the origin of this NFT");
    }

    if (!Object.values(senderAccount.redeemableNFT.creator.requestQueue).includes(nft.id)) {
      throw new Error("NFT is not in origin requesQueue");
    }

    if (nft.redeem.status !== REDEEMSTATUS.REQUESTED) {
      throw new Error("NFT status is not requested");
    }

    nft.redeem.status = REDEEMSTATUS.IDLE;

    nft.journey.unshift({
      activity: JOURNEYACTIVITY.REJECTED,
      date: timestampInSec,
      by: nft.ownerAddress,
      from: senderAddress,
      message: `Requested by ${nft.ownerAddress.toString("hex")} on: ${new Date(nft.redeem.touched).toUTCString()}`,
      amount: 0,
    });

    nft.redeem.touched = timestampInSec;
    senderAccount.redeemableNFT.creator.rejected++;
    const senderRequestIndex = senderAccount.redeemableNFT.creator.requestQueue.findIndex((a) => a.equals(nft.id));
    senderAccount.redeemableNFT.creator.requestQueue.splice(senderRequestIndex, 1);

    const redeemMonitor = await getRedeemMonitor(stateStore);

    const redeemMonitorItemIndex = redeemMonitor.all.findIndex((a) => {
      a.nft.equals(nft.id);
    });
    redeemMonitor.all.splice(redeemMonitorItemIndex, 1);

    if (redeemMonitor.all.length > 0) {
      redeemMonitor.all.sort((a, b) => a - b);
      if (redeemMonitor.checkpoint > redeemMonitor.all[0].time) {
        redeemMonitor.checkpoint = redeemMonitor.all[0].time;
      }
    } else {
      redeemMonitor.checkpoint = 0;
    }

    await setNFTById(stateStore, nft.id, nft);
    await stateStore.account.set(senderAddress, senderAccount);
    await setRedeemMonitor(stateStore, redeemMonitor);
  }
}

module.exports = RejectRedeemAsset;
