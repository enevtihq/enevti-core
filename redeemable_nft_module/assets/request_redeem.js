const { BaseAsset } = require("lisk-sdk");
const { REDEEMABLE_NFT_REQUEST_REDEEM } = require("../constants/id");
const RequestRedeemAssetSchema = require("../schemas/assets/request_redeem");
const REDEEMSTATUS = require("../constants/redeem_status");
const RECURRING = require("../constants/recurring");
const { getNFTById, setNFTById, getRedeemMonitor, setRedeemMonitor } = require("../utils/chain_state");
const FEE = require("../constants/fee");

class RequestRedeemAsset extends BaseAsset {
  name = "requestRedeemAsset";
  id = REDEEMABLE_NFT_REQUEST_REDEEM;
  schema = RequestRedeemAssetSchema;

  async apply({ asset, stateStore, reducerHandler, transaction }) {
    const senderAddress = transaction.senderAddress;
    const senderAccount = await stateStore.account.get(senderAddress);
    const timestampInSec = await stateStore.chain.lastBlockHeaders[0].timestamp;
    const blockDate = new Date(timestampInSec * 1000);
    const nft = await getNFTById(stateStore, asset.nftId);

    if (!nft) {
      throw new Error("NFT doesn't exist");
    }

    if (!nft.ownerAddress.equals(senderAddress)) {
      throw new Error("Sender is not the owner");
    }

    if (nft.redeem.status !== REDEEMSTATUS.IDLE) {
      throw new Error("NFT status is not idle");
    }

    if (nft.redeem.count >= nft.redeem.limit) {
      throw new Error("NFT redeem count exceet limit");
    }

    if (new Date(blockDate).getUTCHours() !== nft.redeem.from.hour) {
      throw new Error("Its not the hour to redeem!");
    }

    if (new Date(blockDate).getUTCMinutes() !== nft.redeem.from.minute) {
      throw new Error("Its not the minute to redeem!");
    }

    let baseTime;
    switch (nft.redeem.recurring) {
      case RECURRING.PERWEEK:
        if (blockDate.getUTCDay() !== nft.redeem.time.day) {
          throw new Error("Its not the day to redeem!");
        }
        baseTime = Date.UTC(blockDate.getUTCFullYear, blockDate.getUTCMonth, blockDate.getUTCDate, nft.redeem.from.hour, nft.redeem.from.minute);
        break;
      case RECURRING.PERMONTH:
        if (blockDate.getUTCDate() !== nft.redeem.time.date) {
          throw new Error("Its not the date to redeem!");
        }
        baseTime = Date.UTC(blockDate.getUTCFullYear, blockDate.getUTCMonth, nft.redeem.time.date, nft.redeem.from.hour, nft.redeem.from.minute);
        break;
      case RECURRING.PERYEAR:
        if (blockDate.getUTCDate() !== nft.redeem.time.date) {
          throw new Error("Its not the date to redeem!");
        }
        if (blockDate.getUTCMonth() !== nft.redeem.time.month) {
          throw new Error("Its not the month to redeem!");
        }
        baseTime = Date.UTC(blockDate.getUTCFullYear, nft.redeem.time.month, nft.redeem.time.date, nft.redeem.from.hour, nft.redeem.from.minute);
        break;
      case RECURRING.ONCE:
        if (blockDate.getUTCDate() !== nft.redeem.time.date) {
          throw new Error("Its not the date to redeem!");
        }
        if (blockDate.getUTCMonth() !== nft.redeem.time.month) {
          throw new Error("Its not the month to redeem!");
        }
        if (blockDate.getUTCFullYear !== nft.redeem.time.year) {
          throw new Error("It's not the year to redeem!");
        }
        baseTime = Date.UTC(nft.redeem.time.year, nft.redeem.time.month, nft.redeem.time.date, nft.redeem.from.hour, nft.redeem.from.minute);
        break;
      default:
        throw new Error("FATAL: Reccuring type unknown");
    }

    const redeemMonitor = await getRedeemMonitor(stateStore);
    const checkpointTime = Math.floor(baseTime / 1000) + nft.redeem.until;
    if (redeemMonitor.checkpoint > checkpointTime) {
      redeemMonitor.checkpoint = checkpointTime;
    }
    redeemMonitor.all.push({
      time: checkpointTime,
      nft: nft.id,
      status: REDEEMSTATUS.REQUESTED,
    });
    redeemMonitor.all.sort((a, b) => a.time - b.time);

    const originAddress = nft.originAddress;
    const originAccount = await stateStore.account.get(originAddress);
    nft.redeem.status = REDEEMSTATUS.REQUESTED;
    nft.redeem.touched = checkpointTime;
    originAccount.redeemableNFT.creator.requestQueue.push(nft.id);

    await reducerHandler.invoke("token:debit", {
      address: senderAddress,
      amount: FEE.REDEEM,
    });

    await reducerHandler.invoke("token:credit", {
      address: originAddress,
      amount: FEE.REDEEM,
    });

    senderAccount.redeemableNFT.collector.redeemed.total++;
    const redeemOriginIndex = senderAccount.redeemableNFT.collector.redeemed.detail.findIndex((a) => {
      a.originAddress.equals(originAddress);
    });
    if (redeemOriginIndex !== -1) {
      senderAccount.redeemableNFT.collector.redeemed.detail[redeemOriginIndex].count++;
    } else {
      senderAccount.redeemableNFT.collector.redeemed.detail.push({
        originAddress: originAddress,
        count: 1,
      });
    }

    await setNFTById(stateStore, nft.id, nft);
    await stateStore.account.set(originAddress, originAccount);
    await stateStore.account.set(senderAddress, senderAccount);
    await setRedeemMonitor(stateStore, redeemMonitor);
  }
}

module.exports = RequestRedeemAsset;
