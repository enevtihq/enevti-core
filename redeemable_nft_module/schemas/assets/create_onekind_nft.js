const CreateOneKindNFTAssetSchema = {
  $id: "enevti/redeemable_nft/createOneKindNFT",
  type: "object",
  required: ["name", "description", "data", "utility", "recurring", "time", "from", "until", "redeemLimit", "royalty", "price", "quantity", "mintingExpire", "timestamp"],
  properties: {
    name: {
      dataType: "string",
      fieldNumber: 1,
    },
    description: {
      dataType: "string",
      fieldNumber: 2,
    },
    data: {
      dataType: "string",
      fieldNumber: 3,
    },
    utility: {
      dataType: "string",
      fieldNumber: 4,
    },
    recurring: {
      dataType: "string",
      fieldNumber: 5,
    },
    time: {
      type: "object",
      required: ["day", "date", "month", "year"],
      fieldNumber: 6,
      properties: {
        day: {
          dataType: "uint32",
          fieldNumber: 1,
        },
        date: {
          dataType: "uint32",
          fieldNumber: 2,
        },
        month: {
          dataType: "uint32",
          fieldNumber: 3,
        },
        year: {
          dataType: "uint32",
          fieldNumber: 4,
        },
      },
    },
    from: {
      dataType: "string",
      fieldNumber: 7,
    },
    until: {
      dataType: "string",
      fieldNumber: 8,
    },
    redeemLimit: {
      dataType: "uint32",
      fieldNumber: 9,
    },
    royalty: {
      type: "object",
      required: ["origin", "staker"],
      fieldNumber: 10,
      properties: {
        origin: {
          dataType: "uint32",
          fieldNumber: 1,
        },
        staker: {
          dataType: "uint32",
          fieldNumber: 2,
        },
      },
    },
    price: {
      type: "object",
      required: ["amount", "currency"],
      fieldNumber: 11,
      properties: {
        amount: {
          dataType: "uint64",
          fieldNumber: 1,
        },
        currency: {
          dataType: "string",
          fieldNumber: 2,
        },
      },
    },
    quantity: {
      dataType: "uint32",
      fieldNumber: 12,
    },
    mintingExpire: {
      dataType: "uint32",
      fieldNumber: 13,
    },
    timestamp: {
      dataType: "uint32",
      fieldNumber: 14,
    },
  },
};

module.exports = CreateOneKindNFTAssetSchema;
