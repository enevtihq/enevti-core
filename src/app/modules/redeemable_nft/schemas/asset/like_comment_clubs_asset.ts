export const likeCommentClubsAssetSchema = {
  $id: 'enevti/redeemableNft/likeCommentClubsAsset',
  title: 'likeCommentClubsAsset transaction asset for redeemableNft module',
  type: 'object',
  required: ['id'],
  properties: {
    id: {
      dataType: 'string',
      fieldNumber: 1,
    },
  },
};
