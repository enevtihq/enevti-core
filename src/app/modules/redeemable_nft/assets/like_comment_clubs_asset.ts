import { BaseAsset, ApplyAssetContext, ValidateAssetContext } from 'lisk-sdk';
import { RedeemableNFTAccountProps } from '../../../../types/core/account/profile';
import { LikeCommentClubsProps } from '../../../../types/core/asset/redeemable_nft/like_comment_clubs_asset';
import { ACTIVITY } from '../constants/activity';
import { likeCommentClubsAssetSchema } from '../schemas/asset/like_comment_clubs_asset';
import { getAccountStats, setAccountStats } from '../utils/account_stats';
import { addActivityEngagement } from '../utils/activity';
import { getCollectionById } from '../utils/collection';
import {
  addCommentClubsLikeById,
  getCommentClubsById,
  setCommentClubsById,
} from '../utils/engagement';
import { getNFTById } from '../utils/redeemable_nft';
import { getBlockTimestamp } from '../utils/transaction';

export class LikeCommentClubsAsset extends BaseAsset {
  public name = 'likeCommentClubs';
  public id = 15;

  // Define schema for asset
  public schema = likeCommentClubsAssetSchema;

  public validate(_input: ValidateAssetContext<LikeCommentClubsProps>): void {
    // Validate your asset
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async apply({
    asset,
    transaction,
    stateStore,
  }: ApplyAssetContext<LikeCommentClubsProps>): Promise<void> {
    const timestamp = getBlockTimestamp(stateStore);
    const comment = await getCommentClubsById(stateStore, asset.id);
    if (!comment) {
      throw new Error('Comment doesnt exists');
    }

    if (comment.type === 'nft') {
      const nft = await getNFTById(stateStore, comment.target.toString('hex'));
      if (!nft) {
        throw new Error('NFT not found while checking authorization');
      }
      if (
        Buffer.compare(nft.creator, transaction.senderAddress) !== 0 &&
        Buffer.compare(nft.owner, transaction.senderAddress) !== 0
      ) {
        throw new Error('You are not authorized to give reply on this NFT clubs');
      }
    } else if (comment.type === 'collection') {
      const collection = await getCollectionById(stateStore, comment.target.toString('hex'));
      if (!collection) {
        throw new Error('Collection not found while checking authorization');
      }
      if (
        Buffer.compare(collection.creator, transaction.senderAddress) !== 0 &&
        collection.stat.owner.findIndex(o => Buffer.compare(o, transaction.senderAddress) === 0) ===
          -1
      ) {
        throw new Error('You are not authorized to give reply on this collection clubs');
      }
    }

    comment.like += 1;
    await addCommentClubsLikeById(stateStore, asset.id, transaction.senderAddress);
    await setCommentClubsById(stateStore, asset.id, comment);

    await addActivityEngagement(stateStore, transaction.senderAddress.toString('hex'), {
      transaction: transaction.id,
      name: ACTIVITY.ENGAGEMENT.LIKECOMMENTCLUBS,
      date: BigInt(timestamp),
      target: comment.id,
    });

    const senderAccount = await stateStore.account.get<RedeemableNFTAccountProps>(
      transaction.senderAddress,
    );
    senderAccount.redeemableNft.likeSent += 1;
    await stateStore.account.set(transaction.senderAddress, senderAccount);

    const accountStats = await getAccountStats(
      stateStore,
      transaction.senderAddress.toString('hex'),
    );
    accountStats.likeSent.commentClubs.unshift(comment.id);
    accountStats.likeSent.total = Object.keys(accountStats.likeSent).reduce(
      (prev, current) =>
        Array.isArray(accountStats.likeSent[current])
          ? prev + (accountStats.likeSent[current] as unknown[]).length
          : prev + 0,
      0,
    );
    await setAccountStats(stateStore, transaction.senderAddress.toString('hex'), accountStats);
  }
}
