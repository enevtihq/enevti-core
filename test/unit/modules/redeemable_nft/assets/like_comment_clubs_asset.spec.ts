import { LikeCommentClubsAsset } from '../../../../../src/app/modules/redeemable_nft/assets/like_comment_clubs_asset';

describe('LikeCommentClubsAsset', () => {
  let transactionAsset: LikeCommentClubsAsset;

  beforeEach(() => {
    transactionAsset = new LikeCommentClubsAsset();
  });

  describe('constructor', () => {
    it('should have valid id', () => {
      expect(transactionAsset.id).toEqual(15);
    });

    it('should have valid name', () => {
      expect(transactionAsset.name).toEqual('likeCommentClubs');
    });

    it('should have valid schema', () => {
      expect(transactionAsset.schema).toMatchSnapshot();
    });
  });

  describe('validate', () => {
    describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
  });

  describe('apply', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
  });
});
