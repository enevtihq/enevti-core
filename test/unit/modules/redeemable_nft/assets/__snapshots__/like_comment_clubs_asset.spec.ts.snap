// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LikeCommentClubsAsset constructor should have valid schema 1`] = `
Object {
  "$id": "enevti/redeemableNft/likeCommentClubsAsset",
  "properties": Object {
    "id": Object {
      "dataType": "string",
      "fieldNumber": 1,
    },
  },
  "required": Array [
    "id",
  ],
  "title": "likeCommentClubsAsset transaction asset for redeemableNft module",
  "type": "object",
}
`;
